endpointComponent/endpoint.c

```diff
@@ -0,0 +1,96 @@
+#include "interfaces.h"
+#include "legato.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <termios.h>
+#include "connectivity/conn_http.h"
+#include "utils/crypto_utils.h"
+#include "utils/serializer.h"
+#include "utils/tryte_byte_conv.h"
+#include "utils/uart_utils.h"
+#include "utils/utils.h"
+
+#define MSG "%s:%s"
+
+COMPONENT_INIT {
+  uint8_t ciphertext[1024] = {0}, iv[16] = {0}, raw_msg[1000] = {0};
+  uint32_t raw_msg_len = 1 + ADDR_LEN + 20, ciphertext_len = 0, msg_len;
+  char tryte_msg[1024] = {0}, msg[1024] = {0}, addr[ADDR_LEN + 1] = ADDRESS, next_addr[ADDR_LEN + 1] = {0};
+  srand(time(NULL));
+
+  // init uart for modem
+  int fd = uart_init();
+  if (fd < 0) {
+    LE_ERROR("Error in initializing UART\n");
+  }
+
+  // create log file
+  if (log_address(next_addr) != 0) {
+    LE_ERROR("log address failed\n");
+  }
+
+  char *response = NULL;
+  time_t timer;
+  char time_str[26];
+  struct tm *tm_info;
+
+  fd_set rset;
+  struct timeval tv;
+  tv.tv_sec = 0;
+  tv.tv_usec = 500;
+  // TODO:change to none blocking I/O descriptor
+  while (true) {
+    FD_ZERO(&rset);
+    FD_SET(fd, &rset);
+    select(fd + 1, &rset, NULL, NULL, &tv);
+
+    if (FD_ISSET(fd, &rset)) {
+      time(&timer);
+      tm_info = localtime(&timer);
+      strftime(time_str, 26, "%Y-%m-%d %H:%M:%S", tm_info);
+      printf("%s\n", time_str);
+      gen_trytes(ADDR_LEN, next_addr);
+
+      response = uart_read(fd);
+      // real transmitted data
+      snprintf((char *)raw_msg, raw_msg_len, MSG, next_addr, response);
+      LE_DEBUG("Raw_msg:%s\n", raw_msg);
+      uint8_t private_key[AES_BLOCK_SIZE * 2] = {0};
+      uint8_t id[IMSI_LEN + 1] = {0};
+
+      if (get_aes_key(private_key) != 0) {
+        LE_ERROR("%s\n", "get aes key error");
+      }
+      // fetch Device_ID (IMSI, len <= 15)
+      if (get_device_id((char *)id) != 0) {
+        LE_ERROR("%s\n", "get device id error");
+      }
+
+      ciphertext_len = ta_encrypt(raw_msg, strlen((char *)raw_msg), ciphertext, 1024, iv, private_key, id);
+      if (ciphertext_len == 0) {
+        LE_ERROR("%s\n", "ta_encrypt msg error");
+      }
+      serialize_msg(iv, ciphertext_len, ciphertext, msg, &msg_len);
+      bytes_to_trytes((const unsigned char *)msg, msg_len, tryte_msg);
+
+      // Init http session. verify: check the server CA cert.
+      char msg_body[1024];
+      gen_tryte_msg(tryte_msg, addr, msg_body);
+      if (send_https_msg(HOST, PORT, API, msg_body, 1024) != HTTP_OK) {
```

Comment:
We should allow user to use other `host` and `port` here

Reply:
Enter by the command line or config file?

Reply:
How about using cli at the first step? 
I think it may be easier to implement

Reply:
btw, Where is `HTTP_OK` defined?

Reply:
Already defined inside defined_error.h.

Reply:
thx